### PR TITLE
sourcekitd: build Swift syntax tree more lazily than collecting parsed tokens.

### PR DESCRIPTION
**Pull request by nkcsgexi**
Before this patch, we have one flag (KeepSyntaxInfo) to turn on two syntax
functionalities of parser: (1) collecting parsed tokens for coloring and
(2) building syntax trees. Since sourcekitd is the only consumer of either of these
functionalities, sourcekitd by default always enables such flag.
However, empirical results show (2) is both heavier and less-frequently
needed than (1). Therefore, separating the flag to two flags makes more
sense, where CollectParsedToken controls (1) and BuildSyntaxTree
controls (2).

CollectingParsedToken is always enabled by sourcekitd because
formatting and syntax-coloring need it; however BuildSyntaxTree should
be explicitly switched on by sourcekitd clients.

resolves: rdar://problem/37483076

### CODE REVIEW

**Comment by nkcsgexi** [comment]
@swift-ci please smoke test

**Comment by nkcsgexi** [comment]
@swift-ci please smoke test

File: include/swift/Basic/LangOptions.h [hunk 1]
```diff
@@ -252,9 +252,11 @@ namespace swift {
     /// This is used to guard preemptive testing for the fix-it.
     bool FixStringToSubstringConversions = false;
 
-    /// Whether to create and keep track of a libSyntax tree associated with
-    /// this source file.
-    bool KeepSyntaxInfoInSourceFile = false;
+    /// Whether collect tokens during parsing for syntax coloring.
+    bool CollectSyntaxToken = false;
```
Comment (rintaro) [inline include/swift/Basic/LangOptions.h]: `CollectParsedTokens` maybe.

File: include/swift/AST/Module.h [hunk 2]
```diff
@@ -888,7 +888,8 @@ class SourceFile final : public FileUnit {
   ASTStage_t ASTStage = Parsing;
 
   SourceFile(ModuleDecl &M, SourceFileKind K, Optional<unsigned> bufferID,
-             ImplicitModuleImportKind ModImpKind, bool KeepSyntaxInfo);
+             ImplicitModuleImportKind ModImpKind, bool KeepTokenSyntax = false,
```
Comment (rintaro) [inline include/swift/AST/Module.h]: `TokenRecorder` is not related to `libSyntax`. I feel `KeepParsedTokens` or such are more relevant.

**Comment by nkcsgexi** [comment]
@swift-ci please smoke test

**Review by rintaro — APPROVED** [review]
Thank you!

**Comment by nkcsgexi** [comment]
@swift-ci Please clean test Linux platform

**Comment by nkcsgexi** [comment]
@swift-ci Please clean test Linux platform

**Pull request merged**